dvc/cache/base.py

```diff
@@ -218,40 +218,22 @@ def _do_link(self, from_info, to_info, link_method):
             "Created '%s': %s -> %s", self.cache_types[0], from_info, to_info,
         )
 
-    def _save_file(self, path_info, tree, hash_info, save_link=True, **kwargs):
+    def _save_file(self, path_info, tree, hash_info, **kwargs):
         assert hash_info
 
         cache_info = self.tree.hash_to_path_info(hash_info.value)
-        if tree == self.tree:
-            if self.changed_cache(hash_info):
+        if self.changed_cache(hash_info):
+            with tree.open(path_info, mode="rb") as fobj:
+                # using our makedirs to create dirs with proper permissions
                 self.makedirs(cache_info.parent)
-                self.move(path_info, cache_info)
+                self.tree.upload_fobj(fobj, cache_info)
```

Comment:
Compared to the previous implementation, we don't move things to cache but rather copy them. This might be a problem in case you want to add a file that can only fit once in your filesystem, but that's rather unusual and you have `--to-remote` now to help you with that. A giant PRO here is that it is much easier to recover from `dvc add` failures (no space, permission problems, etc), as the file/dir is lying right there, intact. Keeping old behavior is not a problem and fits into the arch, so I'll bring that back for now, but will start a discussion elsewhere.

Reply:
On the other hand we were about to look into recovering from such errors and this is a good and intuitive way to do it. We've received multiple reports over this year of dvc badly handling initial `dvc add` when there are problems with space or permissions and that's especially bad with directories, where we might be stuck in a limbo of half the files being in cache and half being in the workspace, recovering from such things is a nightmare. Keeping the safer new behavior after all.

Reply:
sorry not much context on my end, just reading your comments. it means we'll hit some performance issues as well, right? E.g. will it do a full copy even on a FS that supports reflinks?

Reply:
Correct. 1 copy when adding the file, that's it.

Reply:
so, the process itself will be way way longer on these modern file systems. Not sure this is good to be honest.

Reply:
... or any file system where people set symlinks/hardlinks I guess

Reply:
Agreed, it would be better to be able to properly recover. I'm keeping it in this PR for now(in WIP), will tackle transfer now and will then bring the optimization back.

Reply:
Ok, checked the next steps with `transfer`, there will be a bit more steps required to make it merge nicely. So brought back the `move` optimization for now. So the behavior is unchanged.